Update user management guide to explain adding users to security group and changing a phone number (#1389)
alan-turing-institute/data-safe-haven: docs/roles/system_manager/manage_users.md

- Be particularly careful never to use them to log in to any user-accessible VMs (such as the SRDs)
```

> ### {{woman}} {{man}} Adding users to security group(s)

Reviewer: Should we also explain removing users from groups (or modifying groups)?

Author: Perhaps "Managing security group(s)" which can include addition and deletion

Reviewer: :+1:

If it starts to get very large it might be worth breaking it down into addition, deletion, _etc._

> Users that didn't have a `GroupName` specified in the `user_details_template.csv`, will need to be added manually, before they can access an SRE.

Reviewer: This is a useful tip, but I think the instructions here are more general than fixing this one problem.

Maybe something like this would be better as a warning further up the page where editing the csv is discussed? At the moment, it says this is optional, which is true, but it sounds like users are unable to work until they are added to an appropriate group.

- Log into the **SHM primary domain controller** (`DC1-SHM-<SHM ID>`) VM using the login credentials {ref}`stored in Azure Key Vault <roles_system_deployer_shm_remote_desktop>`
- In Server Manager click `Tools > Active Directory Users and Computers`
- Click on `Safe Haven Security Groups`
- Find the group that the user needs to be added to. For each SRE there are the following groups:
    - `SG <SRE ID> Data Administrators` (enables database privileges, not write access to `/data`)
    - `SG <SRE ID> Research Users`
    - `SG <SRE ID> System Administrators`
- Right click on the group and click `Properties`
- Click the `Members` tab and click `Add...`
- Enter a part of the user's name and click `Check Names`
- Select the correct user and click `OK`, then click `OK` again until the window closes
- Open a `Powershell` command window with elevated privileges
- Run `C:\Installation\Run_ADSync.ps1`

### {{iphone}} Changing a users phone number

> Sometimes if a user has provided an incorrect phone number, or has a new phone number that needs to be added, the phone number associated with their account will need to be updated.

Reviewer: ```suggestion
If a user has provided an incorrect phone number, or has a new phone number, the phone number associated with their account will need to be updated.
```

> - Login into the Azure Portal and connect to the correct AAD

Reviewer: ```suggestion
- Login into the Azure Portal and navigate to the correct AAD
```

- Open `Azure Active Directory`
- Click `Users` under `Manage`

> - Search for the user and check the box next to their name, then click `Delete` (this could take a few minutes)

Reviewer: If I recall correctly, this is because we can't change the phone numbers on AAD so we essentially recreate the AAD user from the local AD.

Is that right?
It might be worth adding some text to that effect to explain why this process looks a little odd.

Author: Sure, yes that what @jemrobinson suggested, that the DC1 is the source of truth for user info

Reviewer 2: 1. the user shouldn't need to be deleted in AzureAD, in fact doing this will require the user to re-setup MFA
2. email is the one that can't be changed in AzureAD (once set it's used as a possible UUID for the user)

Author: Ah ok, so we don't want to delete in AzureAD that when changing details.

Shall we say that if email is the thing that needs to be changed, the user should be deleted entirely (and that setting up MFA again will be required)?

Author: See my edit

Reviewer 2: Deleting the user on the DC should also delete them on AzureAD - you should (hopefully) not need to edit users on AzureAD at all.

- Be particularly careful never to use them to log in to any user-accessible VMs (such as the SRDs)
```

(adding_users_manually)=

### {{woman}} {{man}} Adding users to security group(s)

> Users will need to be added to the relevant security group before they can access an SRE. This may have already been done if the `GroupName` was specified in the `user_details_template.csv` (see {ref}`generate_user_csv`), but otherwise can be achieved by the following:

Reviewer: Again, this feels like it is in the wrong place.

This section is about adding a user to a group, for any purpose. I think the warning about users not being able to access the SRE if they are in no group would be better placed with the creating users/users csv. That way we might help people avoid that mistake. If you are reading about it here, you have probably already messed up.

- Be particularly careful never to use them to log in to any user-accessible VMs (such as the SRDs)
```

(adding_users_manually)=

### {{woman}} {{man}} Modifying user SRE access

> Users will need to be added to the relevant security group before they can access an SRE. This will have been done by setting the `GroupName` in the `user_details_template.csv` (see {ref}`generate_user_csv`), but can me modified later by the following:

Reviewer: ```suggestion
Users may have been added to one or more security groups through setting the `GroupName` field in the `user_details_template.csv` (see {ref}`generate_user_csv`). Security Group assignments can also be manually modified```

- Log into the **SHM primary domain controller** (`DC1-SHM-<SHM ID>`) VM using the login credentials {ref}`stored in Azure Key Vault <roles_system_deployer_shm_remote_desktop>`
- In Server Manager click `Tools > Active Directory Users and Computers`
- Click on `Safe Haven Security Groups`
- Find the group that the user needs to be added to. For each SRE there are the following groups:
    - `SG <SRE ID> Data Administrators` (enables database privileges, not write access to `/data`)
    - `SG <SRE ID> Research Users`

>     - `SG <SRE ID> System Administrators`

Reviewer: This is good, is there an explanation in the documents for these groups and their meaning? That feels like a missing piece of the puzzle here.

It might be best to put those definitions in the documentation as a reference and link to it here (and in the section about creating users with the csv).

Author: It doesn't look like there's a good explanation for these groups anywhere. Do we think that based on the "Roles" it should be: 

- Researcher: `SG <SRE ID> Research Users`
- Investigator: `SG <SRE ID> Data Administrators`
- System Manager: `SG <SRE ID> System Administrators`

<img width="318" alt="Screenshot 2023-02-21 at 14 09 33" src="https://user-images.githubusercontent.com/5486164/220368191-6f8faf1b-da33-4f23-80dc-53b9f18490ea.png">

I'm unsure why for example researchers don't have database creation privileges? (assuming I'm correct you need to be in `SG <SRE ID> Data Administrators` for that

Reviewer: I think all of those security groups apply to Researchers and it comes from principle of least privilege.

- Research Users - Probably most Researchers, not special permissions
- Data Administrators - Researchers who can create/modify/delete database ~~tables~~ schemas. Given to a smaller number of Researchers to prevent accidents.
- System Administrators - Researchers with elevated privileges through `sudo`. Rarely used but could be useful in Tier0/1 to let groups manage their own packages.

@jemrobinson, @martintoreilly is that correct?

Reviewer 2: The reason for normal users not to be able to create/delete arbitrary schemas is because some SREs have their input data in database form (and we don't want users to be able to change that). All users should be able to create tables in the `public` schema though.

Author: Ok, I will update this info - I think it should still go on the `System Manager -> Managing Data Safe Haven Users` page, but near the top and referenced later

- Right click on the group and click `Properties`
- Click the `Members` tab
- To add a user click `Add...`
    - Enter a part of the user's name and click `Check Names`
    - Select the correct user and click `OK`, then click `OK` again until the window closes
- To remove a user click on the username of the person and then `Remove`
    - Click `Yes` if you're sure this user should no longer have access to this SRE, then click `OK` again until the window closes
- Open a `Powershell` command window with elevated privileges
- Run `C:\Installation\Run_ADSync.ps1`

### {{iphone}} Edit user details

> The `DC1` is the source of truth for user details. If these details need to be changed, they should be changed in the `DC1` and then synchronised to Azure.

Reviewer: ```suggestion
The `DC1` is the source of truth for user details. If these details need to be changed, they should be changed in the `DC1` and then synchronised to Azure AD.
```

- Click on `Safe Haven Research Users`
- Find the person, right click on them and select `Properties`
- To edit a **phone number**, select the `Telephones` tab and edit the `Mobile` number
    - Click `OK` to save the new number
    - Open a `Powershell` command window with elevated privileges
    - Run `C:\Installation\Run_ADSync.ps1`
- To edit a user's **email** or their **username** (or first name or last name) you'll need to delete the user entirely and recreate them, meaning they'll have to set up their accounts (including MFA) again
    - Find the person, right click on them and click `Delete`
    - Click `OK`
    - Open a `Powershell` command window with elevated privileges
    - Run `C:\Installation\Run_ADSync.ps1`
    - Create a new csv (or edit an existing) one with the correct user details (see {ref}`create_new_users`)
    - Run `C:\Installation\CreateUsers.ps1 <path_to_user_details_file>`
    - Run `C:\Installation\Run_ADSync.ps1`
- You can check the changes you made were successful by logging into the Azure Portal as the AAD admin
    - Open `Azure Active Directory`
    - Click on `Users` under `Manage` and search for the user

>     - Click on the user and then `Edit properties` and confirm your changes propagated to Azure

Reviewer: ```suggestion
    - Click on the user and then `Edit properties` and confirm your changes propagated to Azure AD
```

- Open a `Powershell` command window with elevated privileges
- Run `C:\Installation\Run_ADSync.ps1`

### {{iphone}} Edit user details

> The `DC1` is the source of truth for user details. If these details need to be changed, the user should be deleted in the Azure Portal and re-created from the `DC1`.

Reviewer: Will delegate to @jemrobinson here.

I _think_ user details should sync between AAD and the local AD, so deleting and recreating the user shouldn't often be necessary. I recall one entry being fixed, which it seems was email address and not phone number. That might be the only case where you would make a new user.

Reviewer 2: Yes, the users in AzureAD are synchronised from the local AD on the DC. So all changes (apart from resetting MFA) should be done on the local DC.

Reviewer: An that is unidirectional? _I.e._ no sync from AAD to local AD?

Reviewer 2: There is password hash synchronisation, but I think that's just that when a password is reset the hash is sent to both AzureAD and the local AD, I don't think the password hash is synchronised from AzureAD to the local AD.

The only thing I can think of that's synchronised from AzureAD to the local AD is the AzureAD anchor attribute which identifies which local AD user maps to which AzureAD user.
